src/NLog/Internal/TargetWithFilterChain.cs

```diff
@@ -50,6 +51,115 @@ internal class TargetWithFilterChain
+            public readonly int FileSourceLineNumber;
+
+            /// <summary>
+            /// Serves as a hash function for a particular type.
+            /// </summary>
+            /// <returns>
+            /// A hash code for the current <see cref="T:System.Object"/>.
+            /// </returns>
+            public override int GetHashCode()
+            {
+                return MethodName.GetHashCode() ^ FileSourceName.GetHashCode() ^ FileSourceLineNumber;
+            }
+
+            /// <summary>
+            /// Determines if two objects are equal in value.
+            /// </summary>
+            /// <param name="obj">Other object to compare to.</param>
+            /// <returns>True if objects are equal, false otherwise.</returns>
+            public override bool Equals(object obj)
+            {
+                return obj is CallSiteKey key && Equals(key);
+            }
+
+            /// <summary>
+            /// Determines if two objects of the same type are equal in value.
+            /// </summary>
+            /// <param name="other">Other object to compare to.</param>
+            /// <returns>True if objects are equal, false otherwise.</returns>
+            public bool Equals(CallSiteKey other)
+            {
+                return FileSourceLineNumber == other.FileSourceLineNumber
+                    && string.Equals(FileSourceName, other.FileSourceName, StringComparison.Ordinal)
+                    && string.Equals(MethodName, other.MethodName, StringComparison.Ordinal);
+            }
+        }
+
+        internal bool TryCallSiteClassNameOptimization(StackTraceUsage stackTraceUsage, LogEventInfo logEvent)
+        {
+            if (stackTraceUsage != StackTraceUsage.WithCallSite
+#if !SILVERLIGHT
+             && stackTraceUsage != StackTraceUsage.WithCallSiteSource
+#endif
+               )
+                return false;
+
+            if (logEvent.CallSiteInformation == null)
+                return false;
+
+            if (logEvent.HasStackTrace)
+                return false;
+
+            if (string.IsNullOrEmpty(logEvent.CallSiteInformation.CallerFilePath))
+                return false;
+
+            return true;
+        }
+
+        internal bool TryRememberCallSiteClassName(LogEventInfo logEvent)
+        {
+            if (logEvent.CallSiteInformation == null)
+                return false;
+
+            if (string.IsNullOrEmpty(logEvent.CallSiteInformation.CallerFilePath))
+                return false;
+
+            string className = logEvent.CallSiteInformation.GetCallerClassName(null, true, true, true);
+            if (string.IsNullOrEmpty(className))
+                return false;
+
+            if (_callSiteClassNameCache == null)
+                return false;
+
+            string internClassName = logEvent.LoggerName == className ?
+                logEvent.LoggerName :
+#if !NETSTANDARD1_0
+                string.Intern(className);
```

Comment:
why do we need this? Please add some code comments, as interning is really not necessary is most applications

Reply:
Added comment. But the classname-extractor can perform substring, and other logic. Want to avoid allocating class-name-strings for every logging-location in a source-file.